### PR TITLE
304 add derivates to optimization

### PR DESCRIPTION
**Pull request by markushal**

### CODE REVIEW

**Comment by markushal** [comment]
@joAschauer I added derivative production and H2 storage to the optimization problem, as discussed in the last jour fix. Would you like to have a look? I expanded the unit tests (the ones checking for expected output and expected objective function are still failing). There is also a test streamlit app to inspect the results of the optimization, and to edit input parameters ``app_test_optimize.py``.

File: tests/test_optimize_settings.json [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            }
+        ],
+        "ELY": {
+            "EFF": 0.75,
+            "CAPEX_A": 25,
+            "OPEX_F": 10,
+            "OPEX_O": 0,
+            "CONV": {
+                "H2O-L": 0.3
+            }
```
[the remark this answers is not in the document]
Comment (joAschauer) [reply tests/test_optimize_settings.json:20]: I am not sure if it is save to commit this suggestion directly. Better redo it.

File: flh_opt/api_opt.py [hunk 2]
```diff
@@ -96,16 +105,17 @@ def optimize(input_data: OptInputDataType) -> tuple[OptOutputDataType, Network]:
...
+    carriers = ["ELEC", "H2", "HEAT", "CO2-G", "H2O-L"]
+    for c in carriers:
+        n.add("Bus", name=c, carrier=c)
+        n.add("Carrier", name=c)
 
-    # add carriers:
-    n.add("Carrier", "Electricity")
-    n.add("Carrier", "H2")
-    n.add("Carrier", "H2O")
+    if "DERIV" in input_data.keys():
```
[the remark this answers is not in the document]
Comment (markushal) [reply flh_opt/api_opt.py:114]: In the current implementation, it is optional. You dont have a DERIV process if the final product is H2. But we could also make DERIV a mandatory key for input_data and result_data, and leave it empty if there is no DERIV process. What do you think, @wingechr ?

File: tests/test_opt.py [hunk 3]
```diff
@@ -24,84 +25,47 @@ def rec_approx(x):
...
+def call_optimize(request):
+    input_data = request.param
     [res, n] = optimize(input_data)
+    return [res, n, input_data]
 
-    # write to netcdf file:
-    n.export_to_netcdf("tmp.nc")
 
-    # Test for expected objective value:
+def test_optimize_optimal_solution(call_optimize):
```
[the remark this answers is not in the document]
Comment (markushal) [reply tests/test_opt.py:45]: Yes, i really like these pytest features now :)

File: app_test_optimize.py [hunk 4]
```diff
@@ -83,65 +75,112 @@ def add_vertical_lines(fig: Figure, x_values: list):
...
     )
-    res2.loc["Total", "total cost (USD/MWhH2)"] = res2["total cost (USD/MWhH2)"].sum()
-    st.dataframe(res2.round(1))
-    st.subheader("Aggregate results")
-    st.dataframe(res.round(2))
+    res2.loc["Total", "total cost (USD/MWh final product)"] = res2[
+        "total cost (USD/MWh final product)"
+    ].sum()
+
+    n.export_to_netcdf(f"tests/{input_data['id']}_via_streamlit.nc")
```
[the remark this answers is not in the document]
Comment (markushal) [reply app_test_optimize.py:119]: Hmmm, I cant reproduce your error. The dashboard is very untidy. I would remove it once we are happy with the results. But I would like to keep the export for now, it is nice for quick testing

File: tests/test_optimize_settings.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (markushal) [reply tests/test_optimize_settings.json:39]: Good idead. I fixed this in ef5a3eff6a0b99971986f0cea3747e4e8e9c8195

**Comment by markushal** [comment]
@joAschauer @wingechr Could you fix the issues with ``flh_opt/_types.py`` and api related stuff? From my point of view, we could merge this and feed the real data from the tool into the optimization.

**Review by joAschauer — COMMENTED** [review]
Exporting CH4 network to netcdf fails: 
```python
FAILED tests/test_opt.py::test_optimize_export_to_netcdf[CH4, hybrid] - ValueError: unable to infer dtype on variable 'links_bus3'; object array contains mixed native types: str_, str
```
This is because the "link_bus3" contains a NaN and the NaN cannot be casted to string. 

@markushal is it okay that the link contains NaN?

The tests are not running on github actions because we only run them on pull requests against `main` or `develop`. 

<details>
<summary>Full pytest output</summary>

```
================================================================================================================================== FAILURES ================================================================================================================================== 
________________________________________________________________________________________________________________ test_optimize_export_to_netcdf[CH4, hybrid] _________________________________________________________________________________________________________________ 

call_optimize = [{'DERIV': {'FLH': 0.6110525463600905}, 'ELY': {'FLH': 0.5602974761096946}, 'EL_STR': {'CAP_F': -0.0}, 'H2_STR': {'CAP...X_A': 50, 'EFF': 0.9, 'OPEX_F': 0, 'OPEX_O': 0}, 'H2_STR': {'CAPEX_A': 5, 'EFF': 0.9, 'OPEX_F': 0, 'OPEX_O': 0}, ...}}]

    def test_optimize_export_to_netcdf(call_optimize):
        """Write network to netcdf file."""
        [res, n, input_data] = call_optimize
>       n.export_to_netcdf(f"tests/{input_data['id']}.nc")

tests\test_opt.py:56:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  
.env\lib\site-packages\pypsa\io.py:733: in export_to_netcdf
    with ExporterNetCDF(path, compression, float32) as exporter:
.env\lib\site-packages\pypsa\io.py:72: in __exit__
    self.finish()
.env\lib\site-packages\pypsa\io.py:421: in finish
    self.ds.to_netcdf(self.path)
.env\lib\site-packages\xarray\core\dataset.py:2298: in to_netcdf
    return to_netcdf(  # type: ignore  # mypy cannot resolve the overloads:(
.env\lib\site-packages\xarray\backends\api.py:1339: in to_netcdf
    dump_to_store(
.env\lib\site-packages\xarray\backends\api.py:1386: in dump_to_store
    store.store(variables, attrs, check_encoding, writer, unlimited_dims=unlimited_dims)
.env\lib\site-packages\xarray\backends\common.py:393: in store
    variables, attributes = self.encode(variables, attributes)
.env\lib\site-packages\xarray\backends\common.py:482: in encode
    variables, attributes = cf_encoder(variables, attributes)
.env\lib\site-packages\xarray\conventions.py:795: in cf_encoder
    new_vars = {k: encode_cf_variable(v, name=k) for k, v in variables.items()}
.env\lib\site-packages\xarray\conventions.py:795: in <dictcomp>
    new_vars = {k: encode_cf_variable(v, name=k) for k, v in variables.items()}
.env\lib\site-packages\xarray\conventions.py:199: in encode_cf_variable
    var = ensure_dtype_not_object(var, name=name)
.env\lib\site-packages\xarray\conventions.py:154: in ensure_dtype_not_object
    data = _copy_with_dtype(data, dtype=_infer_dtype(data, name))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

array = array(['nan', 'HEAT'], dtype=object), name = 'links_bus3'

    def _infer_dtype(array, name=None):
        """Given an object array with no missing values, infer its dtype from all elements."""
        if array.dtype.kind != "O":
            raise TypeError("infer_type must be called on a dtype=object array")

        if array.size == 0:
            return np.dtype(float)

        native_dtypes = set(np.vectorize(type, otypes=[object])(array.ravel()))
        if len(native_dtypes) > 1 and native_dtypes != {bytes, str}:
>           raise ValueError(
                "unable to infer dtype on variable {!r}; object array "
                "contains mixed native types: {}".format(
                    name, ", ".join(x.__name__ for x in native_dtypes)
                )
            )
E           ValueError: unable to infer dtype on variable 'links_bus3'; object array contains mixed native types: str_, str

.env\lib\site-packages\xarray\conventions.py:61: ValueError
```

</details>

**Review by wingechr — APPROVED** [review]

**Pull request merged**